Tooling for integration tests

jrcamp opened the pull request:
This adds make and circleci config needed for running separate integration
tests. These tests are intended to still be run from go test so that they can
output coverage data but they may be longer running and talk to external
services like docker.

Note: this is pulling out the build changes from https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/306 to breakup the PRs.

jrcamp commented:
@jchengsfx

jrcamp commented on Makefile:
```diff
@@ -22,26 +28,20 @@ all: common otelcontribcol
-.PHONY: precommit
-precommit:
-	$(MAKE) gotidy
-	$(MAKE) ci
```
I couldn't find a target called `ci` so I removed this. However, it would be nice to have a pre-commit hook, maybe using [pre-commit](https://pre-commit.com/).

jchengsfx commented on .circleci/config.yml:
```diff
@@ -14,8 +14,34 @@ executors:
+  setup:
+    steps:
+      - checkout
+      - restore_module_cache
```
`save_module_cache` is not executed anywhere.

jrcamp replied:
Yeah that's getting addressed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/310.

owais commented on .circleci/config.yml:
```diff
@@ -135,32 +161,29 @@ workflows:
+      - integration-tests:
+          filters:
+            tags:
+              only: /.*/
```
Does this mean it'll run on all branches *and* all tags?

jrcamp replied:
I just copied from unit tests. But yes looks like that's the case.

owais commented on .circleci/config.yml:
```diff
@@ -294,3 +317,24 @@ jobs:
                 }'
           when: on_fail
+
+  integration-tests:
```
Are using the machine executor because we want to be able to run docker? I thought enabling remote docker execution context on docker executor allowed to spawn docker containers from regular executors as well.

jrcamp replied:
Yeah you can spawn docker containers but you can't connect to them directly. See https://circleci.com/docs/2.0/building-docker-images/#accessing-services

> It is not possible to start a service in remote docker and ping it directly from a primary container or to start a primary container that can ping a service in remote docker.

owais commented on Makefile:
```diff
@@ -22,26 +28,20 @@ all: common otelcontribcol
-	    $(GOTEST) $(GOTEST_OPT_WITH_COVERAGE) ./... && \
-	 	go tool cover -html=coverage.txt -o coverage.html ); \
-	done
+	@$(MAKE) for-all CMD="make do-unit-tests-with-cover"
```
- Why do we want to run each module separately now?
- Does this mean we'll need to upload coverage reports from each module to coveralls now?

jrcamp replied:
> * Why do we want to run each module separately now?
It was already being done separately, it's just that the looping was done in bash inside the makefile vs. calling for-all. Thought it'd be better to use for-all and also have the make target available in common so you can run it on an individual module if desired.

> * Does this mean we'll need to upload coverage reports from each module to coveralls now?

I think that was already the case. It already generated a coverage.txt inside each module which codecov collects at the end of circleci run.

owais replied:
I see. For some reason I thought we used to run a single `go test ./...` command to test the entire repo in a single run.

owais commented on Makefile:
```diff
@@ -22,26 +28,20 @@ all: common otelcontribcol
+.PHONY: integration-tests-with-cover
+integration-tests-with-cover:
+	@echo $(INTEGRATION_TEST_MODULES)
+	@$(MAKE) for-all CMD="make do-integration-tests-with-cover" ALL_MODULES="$(INTEGRATION_TEST_MODULES)"
```
I see we've added build tag to the integration test below. Do we need to specify modules like this integration tests? Could we run only tests with integration build flag here and skip the same ones in unit tests?

jrcamp replied:
You mean have a way to automatically detect which modules contain a `// +build integration` directive? It's possible with a regex but seems hacky. I don't know of a simple way to do it and didn't want to spend time to optimize too heavily given it's still early days for integrations tests.

jrcamp replied:
Don't think we need to worry about skipping anything for unit tests since pretty much all modules should have unit tests.

owais replied:
Never mind, you are already doing what I suggested (using --tags). The bash/make code is hard to read for me as must be obvious by now to you :)

owais reviewed (approved):
LGTM

tigrannajaryan commented:
>Merging #321 into master will decrease coverage by 3.54%.

@jrcamp can you clarify this^^^ ?

jrcamp committed:
Tooling for integration tests

This adds make and circleci config needed for running separate integration
tests. These tests are intended to still be run from go test so that they can
output coverage data but they may be longer running and talk to external
services like docker.

jrcamp commented:
@tigrannajaryan not sure, something with codecov was out of sync, rebasing against master seemed to fix it. There's 3 new files listed without coverage because we don't appear to have been running tests for the root module before but now are.
